### PR TITLE
Deprecate blogID in favor of dotComID

### LINKED ISSUE #4618: PostService sends blogID=0 for Jetpack sites connected to a secondary account
In #4275 we changed the remotes to stop passing managed objects and we now pass blog IDs.

In this edge case `blog.blogID = 0`, which wasn't a problem for XML-RPC, but since the site is connected to Jetpack it will use REST.

To reproduce:
- Sign in to the app with a WordPress.com account
- Add a Jetpack site connected to a different WordPress.com account
- Go to stats on the Jetpack site and configure the second WordPress.com account
- Try to refresh posts on the Jetpack site
- Get a `Unknown blog [unknown_blog]` error

/cc @SergioEstevao

### PR DESCRIPTION
**Pull request by koke**
We haven't used blogID for a while for self hosted blogs, as it's pointless: most self hosted will have a blogID=1 if they're not multisite, and XML-RPC requires a blogID parameter on most calls but it's ignored.

So for a while now, we've just stored `0` as the blogID for self-hosted. For wp.com blogs, `blogID` had the actual site ID.

To avoid confusion, I'm deprecating `blogID`. From now on, only `dotComID` should be used.
`dotComID` has the wp.com ID of a WordPress.com or Jetpack site.

For XML-RPC, we're now explicitly sending `0` as the first (blogID) parameter when required.

On a future PR, I'll rename the core data `blogID` attribute to `dotComID`.

## Notes

I don't like this:

```[[BlogServiceRemoteREST alloc] initWithApi:blog.restApi siteID:blog.dotComID]```

Knowing the implementation, I know that if `restApi != nil` then `dotComID` should be set, but it's not enforced at compile time. I'm not sure if it can be.

I've removed a bunch of tests `test*ThrowsExceptionWithoutBlog` because they don't make sense anymore. The blog is provided when initialising the Remote class, and thanks to the new nullability annotations you'll get a warning if you try to pass a `nil` blogID.

I think it's pretty cool how `SiteServiceRemoteREST` disables the parent's `initWithApi:` initialiser by re-declaring it with `__unavailable`. Cool new trick :)

Fixes #4618 

Needs Review: @aerych

### CODE REVIEW

**Comment by koke** [comment]
I put @aerych as the reviewer because I had some doubts about reader comments when refactoring the comment remotes, but @SergioEstevao feel free to take a look and review as well.

File: WordPress/Classes/Models/Blog.m [hunk 1]
```diff
@@ -453,21 +453,23 @@ - (BOOL)jetpackAccountIsDefaultAccount
...
     }
+    [self didAccessValueForKey:@"blogID"];
+    return dotComID;
+}
+
+- (void)setDotComID:(NSNumber *)dotComID
+{
+    [self willChangeValueForKey:@"blogID"];
+    [self setPrimitiveValue:dotComID forKey:@"blogID"];
+    [self didChangeValueForKey:@"blogID"];
```
Comment (aerych) [inline WordPress/Classes/Models/Blog.m]: Maybe make a const for all the instances of `@"blogID"` ?

File: WordPress/Classes/Networking/ServiceRemoteXMLRPC.m [hunk 2]
```diff
@@ -22,16 +22,12 @@ - (id)initWithApi:(WPXMLRPCClient *)api username:(NSString *)username password:(
...
-    NSMutableArray *result = [NSMutableArray array];
-    NSString *password = self.password ?: [NSString string];
-    NSString *username = self.username ?: [NSString string];
-    
-    [result addObject:blogID];
-    [result addObject:username];
-    [result addObject:password];
-    
+- (NSArray *)defaultXMLRPCArguments {
+    return @[@0, self.username, self.password];
```
Comment (aerych) [inline WordPress/Classes/Networking/ServiceRemoteXMLRPC.m]: I know what the 0 is all about, but maybe add a comment for posterity? The reasons might not be obvious to newer devs.

**Comment by aerych** [comment]
Hiya @koke 
Changes look good, I just made a couple of nitpicky comments.  I spent a bit of time testing this branch with a wpcom test site, a self-hosted test site, and a wpcom site with no blogs.  Everything related to the changes in this branch seemed to work as expected.  
:shipit: from me :)

File: WordPress/Classes/Models/Blog.m (hunk 1 above)
Comment (koke) [inline WordPress/Classes/Models/Blog.m]: I'm not sure we'd gain much, considering I'm going to remove blogID once we test this in production a bit more.

Comment (aerych) [inline WordPress/Classes/Models/Blog.m]: Fair enough

**Pull request merged**